README
General information
===================

The Symbion SSL Proxy is a network server application. It listens on a TCP
port, accepts SSL connections, and forwards them to an other (local or remote)
TCP port, or UNIX domain socket.
It is NOT a HTTPS server itself, but it can act as a HTTPS server, if you run
an SSL Proxy server on port 443 which forwards the connections to port 80.
SSL Proxy is tested with HTTP, but it should work with any SSL protocols, I
think.

Security
========
SSL Proxy was developed with security in mind. When it allocated the server
TCP port and loaded everything from the filesystem, it chroot()s to a
(possibly empty) directory and changes the process's real and effective user
id to a specified user.

Install
=======
Make sure you have installed OpenSSL, than run make in the main directory.
Copy ssl_proxy to anywhere you want :)

Key generation
==============
SSL Proxy can not generate keys, use ssleay's req utility to do it:
	# mkdir /etc/symbion; cd /etc/symbion
	# openssl req -days 365 -nodes -new -x509 -out cert.pem -keyout key.pem
	# chmod go-rwx key.pem

Usage
=====
SSL Proxy does not has a config file, it simply has some command line
options.
    General options:
	-h			Usage information.
	-d			Turn on debugging. SSL Proxy will not go into
				the background and it will print a lot of
				debug information to stderr.
	-f			Do not detach from the terminal and run in the
				forground. Useful if you want to run SSL Proxy
				from Symbion Daemon Tool.
	-i			Provide some info to the server about the
				client (see below)
	-s [<server host>:]<server port>
				Specify the port, which SSL Proxy will listen
				on (or address to bind to a specific interface).
	-c [<client host>:]<client port>
				Specify the TCP port, which SSL Proxy will
				connect to as a client (The forwarding port).
	-c unix:<socket path>
				Specify the UNIX domain socket, which SSL Proxy
				will connect to as a client (The forwarding port).
	-m <max connection>	SSL Proxy will accept maximum this number of
				connections.
    SSL options:
	-C <certificate file>	SSL Public Certificate file (see Key generation).
	-K <key file>		SSL Private Key File (see Key generation).
    Security options:
	-u <user/uid>		Change real and effective UID to this after
				initialization.
	-r <chroot dir>		Chroot to the specified directory after
				initialization.
    Buffer size options:
	-U <upward buffer>	The size of the buffer used for
				client -> server data transfer.
	-D <downward buffer>	The size of the buffer used for
				server -> client data transfer.
Defaults:
    # ssl_proxy -s 443 -c localhost:80 -m 32 -C /etc/symbion/cert.pem \
	    -K /etc/symbion/key.pem -U 2048 -D 8192

Notes
=====
SSL Proxy 1.0.0 introduced the ability to connect to UNIX domain sockets, not
just TCP sockets. Please note that if you use UNIX domain sockets and you also
use the -r (chroot) feature to make your system more secure, than the socket
file must reside under the chrooted directory. The path specified in the -c
option must be relative to the chrooted directory.
Also note that if you use UNIX domain sockets with the -u (setuid) feature,
then the user must have read and write permission to the socket file.
The TCP sockets has no similar limitations, because they are not associated
with filesystem objects.

The file provided with the -v option can contain several CA certificates in
PEM format. If you use the -V (certificate directory) option, then each file
have to contain exactly ONE certificate. The files are looked up by the CA
subject name hash value, which must be available. You can create symlinks with
the c_rehash utility, contained in the openssl package.

Client info feature
===================
SSL Proxy can provide client information to the server in a special format.

This is a new feature in version 1.0.7, so I do not know of any server software
that can use this information except my own special purpose software. If I
receive information about any software available that uses this information, I
will publicate it on our website, www.symbion.hu.

The usage is very simple. You provide the -i option to SSL Proxy. When a
client connects, SSL Proxy connects to the server the usual way. But before
the data sent by the client is forwarded to the server, SSL Proxy sends an
information line to the server. This is an example of a HTTPS connection with
the -i option:

-----8<-----
szilu@maia:[~]$ nc -l -p 8080
#@ip=127.0.0.1 port=57223
GET / HTTP/1.1
Host: localhost:8443
User-Agent: Links (2.1pre37; Linux 2.6.27-11-eeepc i686; 80x24)
Accept: */*
Accept-Encoding: gzip, deflate
...
----->8-----

As you can see, an extra line beginning with '#@' is sent to the server. It
contains the IP address and the TCP port of the client and optionally the
common name (cn) of the client, if it has provided a certificate.

This information can be used for example for logging or access control purposes.

Examples
========
If you are running a HTTP server at port 80 which does not has SSL support,
and you want it to work on SSL too, than defaults are good for you:
	# ssl_proxy
If you would like to use maximal security level, you can use:
	# mkdir /etc/symbion/chroot_dir
	# ssl_proxy -u nobody -r /etc/symbion/chroot_dir

How to report bugs
==================
To report a bug, send mail to sslproxy-users@lists.sourceforge.net.
In the mail include:

* The version

* Information about your system. For instance:

    - What operating system and version
    - What version of OpenSSL
    - What version of the C library

  And anything else you think is relevant.

* How to reproduce the bug. 

* The text that was printed out (Debug information).

You can also use tha SourceForge bugtracking system at
    http://sourceforge.net/tracker/?group_id=21298

Patches
=======

Patches can be sent to tha sslproxy-users@lists.sourceforge.net mailing list.
Please include your name and email address.

If the patch fixes a bug, it is usually a good idea to include
all the information described in "How to Report Bugs".

    Szilard Hajba <szilu@symbion.hu>


ChangeLog
2009-02-27 15:03:46    Szilard Hajba <szilu@symbion.hu>

	* An other client info bugfix
	* Release 1.1.2

2009-02-27 13:43:34    Szilard Hajba <szilu@symbion.hu>   

	* Client info bugfixes
	* Release 1.1.1

2009-02-13 18:26:27    Szilard Hajba <szilu@symbion.hu>

	* Client info option
	* Release 1.1.0

2008-06-23 14:13:57    Szilard Hajba <szilu@symbion.hu>

	* Client certificate validation
	* Private release 1.0.7

2007-12-13 22:49:48    Szilard Hajba <szilu@symbion.hu>

	* Repository moved to GIT

2006-08-16 11:30:36    Szilard Hajba <szilu@symbion.hu>

	* Some cleanup in the client side connection code
	* Type conversion to compile cleanly on OS X
	* Private release 1.0.6

2005-09-28 17:58:31    Szilard Hajba <szilu@symbion.hu>

	* Improved certificate handling (chained certificates are now supported)
	* Powered by logo
	* Release 1.0.5

2005-08-07 09:19:53    Szilard Hajba <szilu@symbion.hu>

	* More error information on SSL protocol errors during SSL_accept()
	* README: Report bugs to the mailing list
	* Added -U, -D options (buffer size)
	* Release 1.0.4

2005-04-08 11:08:45    Szilard Hajba <szilu@symbion.hu>

	* Some new compiler warnings fixed
	* Connection handling cleanup
	* Release 1.0.3

2005-04-07 17:57:23    Szilard Hajba <szilu@symbion.hu>

	* #1178558 fixed

2004-04-13 21:20:41    Szilard Hajba <szilu@symbion.hu>

	* Added -f option (run in the foreground)

2004-03-02 10:37:45    Szilard Hajba <szilu@symbion.hu>

	* BUGFIX: 100% CPU usage during ssl_accept state

2003-09-23 19:49:45    Szilard Hajba <szilu@symbion.hu>

	* Listen on multiple interfaces (patch by Marco Lanzotti)
	* Release 1.0.1

2001-11-05 03:26:26    Szilard Hajba <szilu@revai.hu>

	* UNIX domain socket used Linux specific headers, now it compiles on
		FreeBSD.

2001-11-04 23:51:49    Szilard Hajba <szilu@revai.hu>

	* Support for UNIX domain sockets (-c unix:<path>)
	* Log errors to syslog
	* Stable release 1.0.0

2001-10-08 01:39:46    Szilard Hajba <szilu@revai.hu>

	* Some bugs were fixed, now it is tested with large file transfers

2001-06-14 17:44:03    Szilard Hajba <szilu@revai.hu>

	* Write buffering bug fix. The bug caused some data damage when large
		files transferred with HTTPS.
	* Two config options added for the two way buffer sizes.

2001-02-16 23:47:29    Szilard Hajba <szilu@revai.hu>

	* Initial release, ChangeLog started


ssl_proxy.c
/* Symbion SSL Proxy
 * Copyright (C) 2000-2005 Szilard Hajba
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

//#define VERSION "1.0.7"
#define MAX_CONNECTION 32
//#define CS_BUFFER_LEN 2
//#define SC_BUFFER_LEN 40
#define CS_BUFFER_LEN 2048
#define SC_BUFFER_LEN 8192
#define PEM_DIR "/etc/symbion"
#define CERT_FILE "cert.pem"
#define KEY_FILE "key.pem"
#define SLEEP_US 50000

#include <stdio.h>
#include <stdarg.h>
#include <syslog.h>
#include <unistd.h>
#include <signal.h>
#include <netinet/in.h>
#include <sys/un.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <fcntl.h>
#include <string.h>
#include <pwd.h>
#include <arpa/inet.h>

#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/x509.h>
#include <openssl/pem.h>
#include <openssl/crypto.h>
#include <openssl/rand.h>

int debug_flag=0;
int fg_flag=0;
int info_flag=0;
int max_conn=MAX_CONNECTION;
int cs_buflen=CS_BUFFER_LEN, sc_buflen=SC_BUFFER_LEN;
char *server_addr="0.0.0.0";
int server_port=443;
char *client_addr="localhost";
int client_port=80;
char *cert_file=PEM_DIR"/"CERT_FILE, *key_file=PEM_DIR"/"KEY_FILE;
char *chroot_dir=NULL, *set_uid=NULL;
char *verify_ca_file=NULL, *verify_ca_dir=NULL;
struct passwd *pass;

int server_socket;
SSL_CTX *server_ssl_ctx;
//X509 *ssl_public_cert;
//RSA *ssl_private_key;

int client_s_family=AF_INET;
struct sockaddr *client_sa;
struct sockaddr_in client_sa_in;
struct sockaddr_un client_sa_un;
int client_sa_len;

typedef enum {cs_disconnected, cs_accept, cs_connecting, cs_connected, cs_closing} ConnStatus;
typedef struct {
    ConnStatus stat;			// Status of the connection
    int server_sock;			// Server side socket id
    struct sockaddr_in server_sa;	// Server's socket address
    int server_sa_len;			// ^^^^^^^^^^^^^^^^^^^^^^^'s len
    SSL *ssl_conn;			// SSL connection structure pointer
    int client_sock;			// Client side socket id
    char *csbuf;			// Server side write buffer
    char *csbuf_b;			// Server side write buffer begin ptr
    char *csbuf_e;			// Server side write buffer end ptr
//    int c_end_req;			// Client requested connection close
    char *scbuf;			// Client side write buffer
    char *scbuf_b;			// Client side write buffer begin ptr
    char *scbuf_e;			// Client side write buffer end ptr
//    int s_end_req;			// Server requested connection close
} Conn;
Conn *conn=NULL;

void conn_close_client(Conn *conn);
void conn_close_server(Conn *conn);

void debug(char *format,...) {
    if (debug_flag) {
	va_list args;
	va_start(args, format);
	vfprintf(stderr, format, args);
	putc('\n', stderr);
	va_end(args);
    }
}

void plog(int level, const char *cls, const char *format,...) {
    char str[8192];
    va_list args;
    va_start(args, format);
    vsprintf(str, format, args);
    va_end(args);
    syslog(level, "%.256s: %.256s\n", cls, str);
    if (debug_flag) debug("LOG: %.256s: %.256s", cls, str);
}

void _sleep()
{
    struct timeval tv={0, SLEEP_US};
    select(0, NULL, NULL, NULL, &tv);
}

// ============================================== Server
int server_init(char *addr, int port, int maxconn) {
    struct sockaddr_in server;
    long ipaddr;

    server_socket=socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket<0) {
	perror("socket()");
	exit(1);
    }
    server.sin_family=AF_INET;
    inet_pton(AF_INET, addr, &ipaddr);
    server.sin_addr.s_addr=ipaddr;
//    server.sin_addr.s_addr=htons(INADDR_ANY);
    server.sin_port=htons(port);
    if (bind(server_socket, (struct sockaddr *)&server, sizeof(server)) < 0) {
	perror("bind()");
	exit(1);
    }
    listen(server_socket, maxconn);
    fcntl(server_socket, F_SETFL, O_NONBLOCK);
    return server_socket;
}

void server_done(void) {
    int ci;
    shutdown(server_socket, 2);
    _sleep();
    close(server_socket);
    for (ci=0; ci<max_conn; ci++)
	if (conn[ci].stat==cs_accept && conn[ci].stat==cs_connected) {
	    conn_close_client(&conn[ci]);
	    conn_close_server(&conn[ci]);
	}
}

// ============================================== Server SSL
static RSA *tmp_rsa_cb(SSL *ssl, int export, int key_len) {
    static RSA *rsa=NULL; 
    debug("Generating new RSA key.. (ex=%d, kl=%d)", export, key_len);
    if (export) {
	rsa=RSA_generate_key(key_len, RSA_F4, NULL, NULL);
    } else {
	plog(LOG_ERR, "tmp_rsa_callback()", "Export not set");
    }
    return rsa;
}

static int verify_callback(int preverify_ok, X509_STORE_CTX *ctx)
{
    fprintf(stderr, "preverify: %d\n", preverify_ok);
    return preverify_ok;
}

void server_ssl_init(void) {
//    FILE *f;
    SSLeay_add_ssl_algorithms();
    SSL_load_error_strings();
    server_ssl_ctx=SSL_CTX_new(SSLv23_server_method());
    SSL_CTX_set_cipher_list(server_ssl_ctx, "HIGH");
    if (!SSL_CTX_set_default_verify_paths(server_ssl_ctx))  {
	fprintf(stderr, "cannot set default path\n");
	exit(1);
    }

    if (!SSL_CTX_use_certificate_chain_file(server_ssl_ctx, cert_file)) {
	fprintf(stderr,"error reading certificate: %.256s\n",
		ERR_error_string(ERR_get_error(), NULL));
	exit(1);
    }
    if (!SSL_CTX_use_PrivateKey_file(server_ssl_ctx, key_file, SSL_FILETYPE_PEM)) {
	fprintf(stderr,"error reading private key: %.256s\n",
		ERR_error_string(ERR_get_error(), NULL));
	exit(1);
    }
    SSL_CTX_set_tmp_rsa_callback(server_ssl_ctx, tmp_rsa_cb);

    if (verify_ca_file || verify_ca_dir) {
/*
	STACK_OF(X509_NAME) *certs;
	certs=SSL_load_client_CA_file(verify_ca_file);
	if (certs) {
	    SSL_CTX_set_client_CA_list(server_ssl_ctx, certs);
	} else {
	    fprintf(stderr,"error reading client CA list: %.256s\n",
		    ERR_error_string(ERR_get_error(), NULL));
	    exit(1);
	}
*/
	SSL_CTX_load_verify_locations(server_ssl_ctx, verify_ca_file, verify_ca_dir);
	SSL_CTX_set_verify(server_ssl_ctx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT | SSL_VERIFY_CLIENT_ONCE, verify_callback);
    }

//    SSL_CTX_set_session_cache_mode(server_ssl_ctx, SSL_SESS_CACHE_OFF);
}

// ============================================== CLient
void client_init(char *addr, int port) {
    if (port) { // TCP connection
	struct hostent *hp;
	client_sa_in.sin_family=AF_INET;
	hp=gethostbyname(addr);
	if (!hp) {
	    perror("gethostbyname()");
	    exit(1);
	}
	bcopy(hp->h_addr, &client_sa_in.sin_addr, hp->h_length);
	client_sa_in.sin_port=htons(port);
	client_sa=(struct sockaddr *)&client_sa_in;
	client_sa_len=sizeof(client_sa_in);
    } else { // UNIX domain socket
	client_sa_un.sun_family=AF_UNIX;
	if (addr) {
	    if (strlen(addr)>=sizeof(client_sa_un.sun_path)) {
		fprintf(stderr, "client_init(): client address too long (allowed: %d)\n",
			(int)sizeof(client_sa_un.sun_path));
		exit(1);
	    } else strcpy(client_sa_un.sun_path, addr);
	} else {
	    fprintf(stderr, "client_init(): client address missing\n");
	    exit(1);
	}
	client_sa=(struct sockaddr *)&client_sa_un;
	client_sa_len=sizeof(client_sa_un);
    }
}

// ============================================== Connection
struct sockaddr_in server_sa;
unsigned int server_sa_len;
int conn_accept(void) {
    int i;
    // Initialize SSL connection (server side)
    int s=accept(server_socket, (struct sockaddr *)&server_sa, &server_sa_len);
    if (s<=0) return 0;
    debug("conn_accept(): Client connected");
    for (i=0; i<max_conn && conn[i].stat!=cs_disconnected; i++);
    if (i==max_conn) {
	plog(LOG_ERR, "accept()", "Internal error");
	close(s);
	return 0;
    }
    debug("accept(): sn=%d sock=%d", i, s);
    conn[i].server_sock=s;
    bcopy(&server_sa, &conn[i].server_sa, server_sa_len);
    conn[i].server_sa_len=server_sa_len;
    conn[i].ssl_conn=SSL_new(server_ssl_ctx);
    SSL_set_fd(conn[i].ssl_conn, conn[i].server_sock);
//    if (!SSL_use_RSAPrivateKey(conn[i].ssl_conn, ssl_private_key)) {
//	plog(LOG_ERR, "accept()", "Error reading private key");
//	SSL_free(conn[i].ssl_conn);
//	close(conn[i].server_sock);
//	return -1;
//    }
//    if (!SSL_use_certificate(conn[i].ssl_conn, ssl_public_cert)) {
//	plog(LOG_ERR, "accept()", "Error reading public certificate");
//	SSL_free(conn[i].ssl_conn);
//	close(conn[i].server_sock);
//	return -1;
//    }
//    SSL_set_verify(conn[i].ssl_conn, 0, NULL);
    BIO_set_nbio(SSL_get_rbio(conn[i].ssl_conn), 0);
    BIO_set_nbio(SSL_get_wbio(conn[i].ssl_conn), 0);
    fcntl(conn[i].server_sock, F_SETFL, O_NONBLOCK);
    conn[i].stat=cs_accept;
    conn[i].scbuf_b=conn[i].scbuf; conn[i].scbuf_e=conn[i].scbuf;
    conn[i].csbuf_b=conn[i].csbuf; conn[i].csbuf_e=conn[i].csbuf;
    return conn[i].server_sock;
}

int conn_ssl_accept(Conn *conn) {
    int ret=SSL_accept(conn->ssl_conn);
//    debug("SSL_accept: %d, SSL_want=%d", ret, SSL_want(conn->ssl_conn));
    if (ret<=0) {
	unsigned long err=SSL_get_error(conn->ssl_conn, ret);
//	debug("SSL_accept: error:%d", err);
	if (err==SSL_ERROR_WANT_READ || err==SSL_ERROR_WANT_WRITE) {
	    return 1;
	} else {
	    plog(LOG_ERR, "accept()", "Accept failed: %.256s", ERR_error_string(err, NULL));
	    ERR_print_errors_fp(stderr);
	}

//	if ((err=ERR_get_error())) {
//	    plog(LOG_ERR, "accept()", "Access failed: %.256s", ERR_error_string(err, NULL));
//	} else if (SSL_want(conn->ssl_conn)>1) return 0;;
	debug("SSL_accept: disconnected.");
	SSL_free(conn->ssl_conn);
	close(conn->server_sock);
	conn->server_sock=conn->client_sock=0;
	conn->stat=cs_disconnected;
	return -1;
/*
    } else if (ret==1) {
	// Successfull negotiation
	X509 *peer;
	if (!(peer=SSL_get_peer_certificate(conn->ssl_conn))
		|| SSL_get_verify_result(conn->ssl_conn)!=X509_V_OK) {
	    unsigned long err=SSL_get_error(conn->ssl_conn, ret);
	    plog(LOG_ERR, "accept()", "SSL handshake: %.256s", ERR_error_string(err, NULL));
	    ERR_print_errors_fp(stderr);
	    debug("SSL_accept: disconnected.");
	    SSL_free(conn->ssl_conn);
	    close(conn->server_sock);
	    conn->stat=cs_disconnected;
	    return -1;
	}
*/
    }

    // Connect to server (client side)
    conn->client_sock=socket(client_s_family, SOCK_STREAM, 0);
    if (conn->client_sock<0) {
	plog(LOG_ERR, "socket()", strerror(errno));
	SSL_free(conn->ssl_conn);
	close(conn->server_sock);
	conn->server_sock=conn->client_sock=0;
	conn->stat=cs_disconnected;
	return -1;
    }
    fcntl(conn->client_sock, F_SETFL, O_NONBLOCK);
    conn->stat=cs_connecting;
/*
    ret=connect(conn->client_sock, client_sa, client_sa_len);
    if (ret<0) {
	if (errno==EINPROGRESS) return 0;
	plog(LOG_ERR, "connect()", strerror(errno));
	close(conn->client_sock);
	SSL_free(conn->ssl_conn);
	close(conn->server_sock);
	conn->stat=cs_disconnected;
	return -1;
    } else {
	conn->stat=cs_connected;
	debug("Connection negotiated");
	if (info_flag) {
	    conn->csbuf_e+=snprintf(conn->csbuf_b, cs_buflen, "#@ip=%s \r\n", "1.1.1.1");
	    debug("INFO: %s", conn->csbuf);
	}
    }
*/
    return 0;
}

/*
void conn_close(Conn *conn) {
    debug("conn_close(): s=%d", conn->server_sock);
    shutdown(conn->client_sock, 2);
    close(conn->client_sock);
    SSL_free(conn->ssl_conn);
    shutdown(conn->server_sock, 2);
    close(conn->server_sock);
    conn->stat=cs_disconnected;
}
*/

void conn_close_client(Conn *conn) {
    debug("conn_close_client(): s=%d", conn->client_sock);
    shutdown(conn->client_sock, 2);
    close(conn->client_sock);
    conn->client_sock=-1;
    if (conn->server_sock==-1) conn->stat=cs_disconnected;
}

void conn_close_server(Conn *conn) {
    debug("conn_close_server(): s=%d", conn->server_sock);
    SSL_free(conn->ssl_conn);
    shutdown(conn->server_sock, 2);
    close(conn->server_sock);
    conn->server_sock=-1;
    if (conn->client_sock==-1) conn->stat=cs_disconnected;
}

void sighandler(int signum) {
    switch (signum) {
	case SIGINT:
	case SIGTERM:
	    plog(LOG_NOTICE, "signal", "Interrupt/terminate");
	    server_done();
	    // If it's possible to remove pid file, try it..
	    // It's not guaranteed to succeed, because of setuid
	    if (!chroot_dir) unlink("/var/run/ssl_proxy.pid");
	    exit(0);
	default:;
    }
}

int main(int argc, char **argv) {
    FILE *pidfile;
    int c, pid, i;
    char *p1, *p2;

    while ((c=getopt(argc, argv, "hdfim:s:c:C:K:u:r:v:V:U:D:")) != EOF)
	switch (c) {
	    case 'h':
		fprintf(stderr, "Symbion SSL proxy " VERSION "\n"
			"usage: %.256s [-d] [-f] [-i] [-s <listen address>] [-c <client address>]\n"
			"              [-m <max connection>] [-C <certificate file>] [-K <key file>]\n"
			"              [-u <user/uid>] [-r <chroot dir>]\n"
			"              [-v <trusted CA file>] [-V <trusted CA dir>]\n"
			"              [-U <upward buffer (default 2048)>] [-D <downward buffer (default 8192)>]\n"
			"        <lister address> = [<host>:]<port>\n"
			"        <client address> = [<host>:]<port> | unix:<path>\n", argv[0]);
		fprintf(stderr, "       %.256s -h\n", argv[0]);
		exit(0);
	    case 'd':
		debug_flag=1;
		break;
	    case 'f':
		fg_flag=1;
		break;
	    case 'i':
		info_flag=1;
		break;
	    case 'm':
		max_conn=atoi(optarg);
		break;
	    case 's':
		server_port=atoi(optarg);
		p1=strtok(optarg, ":");
		p2=strtok(NULL, "");
		if (p2) {
		    server_addr=p1;
		    server_port=atoi(p2);
		} else {
		    server_addr="0.0.0.0"; server_port=atoi(p1);
		}
		break;
	    case 'c':
		p1=strtok(optarg, ":");
		p2=strtok(NULL, "");
		if (p2) {
		    if (!strcmp(p1, "unix")) {
			client_s_family=AF_UNIX;
			client_addr=p2;
			client_port=0;
		    } else {
			client_addr=p1;
			client_port=atoi(p2);
		    }
		} else {
		    client_addr="localhost"; client_port=atoi(p1);
		}
		break;
	    case 'C':
		cert_file=optarg;
		break;
	    case 'K':
		key_file=optarg;
		break;
	    case 'u':
		set_uid=optarg;
		break;
	    case 'r':
		chroot_dir=optarg;
		break;
	    case 'v':
		verify_ca_file=optarg;
		break;
	    case 'V':
		verify_ca_dir=optarg;
		break;
	    case 'U':
		cs_buflen=atoi(optarg);
		break;
	    case 'D':
		sc_buflen=atoi(optarg);
		break;
	}
    debug("Symbion SSL proxy " VERSION);
    signal(SIGINT, sighandler);
    signal(SIGTERM, sighandler);
    signal(SIGPIPE, SIG_IGN);
    // This must be done before process_security_init(), because server_port
    // can be a privileged port, ssl_init use files from the filesystem.
    if (client_s_family==AF_INET)
	debug("Using server: family=INET host=%.256s port=%d", client_addr, client_port);
    else
	debug("Using server: family=UNIX path=%.256s", client_addr);
    server_init(server_addr, server_port, max_conn);
    server_ssl_init();
    client_init(client_addr, client_port);
    if (!debug_flag && !fg_flag && (pid=fork())) {
	pidfile=fopen("/var/run/ssl_proxy.pid", "w");
	if (pidfile) {
	    fprintf(pidfile, "%d\n", pid);
	    fclose(pidfile);
	}
	exit(0);
    }

    // Process security
    if (set_uid) {
	if (!(pass=getpwnam(set_uid))) {perror("getpwnam()"); exit(1);}
    }
    if (chroot_dir) {
	debug("Changing root directory to '%.256s'..", chroot_dir);
	if (chroot(chroot_dir)<0) {perror("chroot()"); exit(1);}
	chdir("/");
    }
    if (set_uid) {
	debug("Changing userID to %.256s..", set_uid);
	setgid(pass->pw_gid);
	setuid(pass->pw_uid);
    }

    conn=malloc(max_conn*sizeof(Conn));
    bzero(conn, max_conn*sizeof(Conn));
    for (i=0; i<max_conn; i++) {
	Conn *c=&conn[i];
	c->scbuf=malloc(sc_buflen);
	c->scbuf_b=c->scbuf; c->scbuf_e=c->scbuf;
	c->csbuf=malloc(cs_buflen);
	c->csbuf_b=c->csbuf; c->csbuf_e=c->csbuf;
    }

    openlog("sslproxy", LOG_PID, LOG_DAEMON);
    if (client_s_family==AF_INET)
	plog(LOG_NOTICE, "init", "version " VERSION " started (family=INET host=%.256s port=%d).",
		client_addr, client_port);
    else
	plog(LOG_NOTICE, "init", "version " VERSION " started (family=UNIX path=%.256s).",
		client_addr);

    while (1) {
	int event=0, ci;
	// Check for incoming connections
	if ((i=conn_accept())>0) {
	    debug("Client connected");
	    event=1;
	}
	for (ci=0; ci<max_conn; ci++) {
	    Conn *cn=&conn[ci];
	    int l;
	    switch (cn->stat) {
		case cs_accept:
		    i=conn_ssl_accept(cn);
//		    cn->c_end_req=0; cn->s_end_req=0;
		    event|=(i==0);
		    break;
		case cs_connecting:
		    if (connect(cn->client_sock, client_sa, client_sa_len)<0) {
			if (errno==EINPROGRESS) break;
//			if (errno==EALREADY) break;
perror("connecting()");
			plog(LOG_ERR, "connect()", strerror(errno));
			close(cn->client_sock);
			SSL_free(cn->ssl_conn);
			close(cn->server_sock);
			cn->stat=cs_disconnected;
		    } else {
			cn->stat=cs_connected;
			debug("Connection negotiated");
			if (info_flag) {
			    struct sockaddr_in client_addr;
			    unsigned int client_addr_len=sizeof(client_addr);
			    X509 *cert;
			    X509_NAME *xn=NULL;
			    char peer_cn[256]="";
			    getpeername(cn->server_sock,
				    (struct sockaddr *)&client_addr,
				    &client_addr_len);
			    cert=SSL_get_peer_certificate(cn->ssl_conn);
			    if (cert) {
				xn=X509_get_subject_name(cert);
				X509_NAME_get_text_by_NID(xn, NID_commonName, peer_cn, 256);
			    }
			    cn->csbuf_e+=snprintf(cn->csbuf_b, cs_buflen,
				    "#@ip=%s port=%d%s%s%s\r\n",
				    inet_ntoa(client_addr.sin_addr),
				    htons(client_addr.sin_port), xn?" cn='":"", peer_cn, xn?"'":"");
			    debug("INFO: %p %d %s", cn, cn->server_sock, cn->csbuf);
			}
		    }
		    break;
		case cs_connected:
		    // Check if data is available on client side
		    if ((l=cs_buflen-(cn->csbuf_e-cn->csbuf))) {
			i=SSL_read(cn->ssl_conn, cn->csbuf_e, l);
			if (!i) { // End of connection
			    debug("Close request: %d", ci);
			    cn->stat=cs_closing; event=1;
//			    cn->c_end_req=1;
			} else if (i<0) { // Error
			    if (errno!=EAGAIN) {
				debug("SSL_read() errno: %d", errno);
				cn->stat=cs_closing; event=1;
//				cn->c_end_req=1;
			    }
			} else cn->csbuf_e+=i;
		    }
		case cs_closing:
		    // Send buffered data to server
		    if ((l=cn->csbuf_e-cn->csbuf_b)>0) {
			i=write(cn->client_sock, cn->csbuf_b, l); event=1;
			if (debug_flag) write(2, cn->csbuf_b, l);
			if (i>=0) {
			    cn->csbuf_b+=i;
			} else {
			    if (errno!=EAGAIN) {
				debug("write() errno: %d", errno);
				cn->csbuf_b=cn->csbuf_e=cn->csbuf;
				cn->stat=cs_closing;
			    }
			}
			if (cn->csbuf_b==cn->csbuf_e) {
			    cn->csbuf_b=cn->csbuf_e=cn->csbuf;
//			    if (cn->c_end_req) conn_close(cn);
			}
		    }
		    if (cn->stat==cs_closing && cn->csbuf_e==cn->csbuf_b) conn_close_client(cn);
		default:;
	    }
	    if (cn->stat==cs_connected || cn->stat==cs_closing) {
		// Check if data is available on server side
		if ((l=sc_buflen-(cn->scbuf_e-cn->scbuf)) && cn->client_sock>=0) {
		    i=read(cn->client_sock, cn->scbuf_e, l);
		    if (!i) { // End of connection
			cn->stat=cs_closing; event=1;
//			cn->s_end_req=1;
		    } else if (i<0) { // Error
			if (errno!=EAGAIN) {
			    debug("read errno: %d", errno);
			    cn->stat=cs_closing; event=1;
//			    cn->s_end_req=1;
			}
		    } else cn->scbuf_e+=i;
		}
		// Send buffered data to client
		if ((l=cn->scbuf_e-cn->scbuf_b)>0 && cn->server_sock>=0) {
		    i=SSL_write(cn->ssl_conn, cn->scbuf_b, l);
		    if (i>0) debug("transfer: buf=%d, b=%d, l=%d, i=%d", cn->scbuf,
			    cn->scbuf_b, l, i);
		    if (i>=0) {
			cn->scbuf_b+=i; event=1;
		    } else if (errno!=EAGAIN) {
			debug("SSL_write() errno: %d", errno);
			cn->scbuf_b=cn->scbuf_e=cn->scbuf;
			event=1;
		    }
		    if (cn->scbuf_b==cn->scbuf_e) {
			cn->scbuf_b=cn->scbuf_e=cn->scbuf;
//			if (cn->s_end_req) conn_close(cn);
		    }
		}
		if (cn->stat==cs_closing && cn->scbuf_e==cn->scbuf_b) conn_close_server(cn);
	    }
	}
	if (!event) _sleep();
    }
    return 0;
}
